collectors/30/kafka_jmx.py

```diff
@@ -1,219 +1,76 @@
 
+# todo(mdodsworth): perhaps we should allow these to be passed through
+#                   having them be static smells funny
 IGNORED_METRICS = frozenset(["Loggers", "MBeanName"])
 
+# todo: figure out if the following variables are set/fetched by tcollector
 # How many times, maximum, will we attempt to restart the JMX collector.
 # If we reach this limit, we'll exit with an error.
 MAX_RESTARTS = 10
-
 TOP = False  # Set to True when we want to terminate.
 RETVAL = 0    # Return value set by signal handler.
 
+class KafkaJmxMonitor(JmxMonitor):
+    USER = "kafka"
+    PROCESS_NAME = "kafka.Kafka"
+
+    def __init__(self, pid, cmd):
+        super(KafkaJmxMonitor, self).__init__(pid,  cmd, MONITORED_MBEANS, self.PROCESS_NAME)
+
+    # Override
+    def process_metric(self, timestamp, metric, value, mbean):
+        jmx_service = ""
+        metric, tags = self.group_metrics(metric)
+
+        # mbean is of the form "domain:key=value,...,foo=bar"
+        # some tags can have spaces, so we need to fix that.
+        mbean_domain, mbean_properties = mbean.rstrip().replace(" ", "_").split(":", 1)
+        mbean_domain = mbean_domain.rstrip().replace("\"", "")
+        mbean_properties = mbean_properties.rstrip().replace("\"", "")
+
+        if mbean_domain not in ("kafka.server",
+                                "kafka.cluster",
+                                "kafka.controller",
+                                "kafka.network",
+                                "kafka.log",
+                                "kafka.consumer",
+                                "java.lang"):
+            utils.err("Unexpected mbean domain = %r" % mbean_domain)
+            return
+
+        mbean_properties = dict(prop.split("=", 1)
+                                for prop in mbean_properties.split(","))
+
+        if mbean_domain == "java.lang":
+            jmx_service = mbean_properties.pop("type", "jvm")
+        elif mbean_domain.startswith("kafka."):
+            jmx_service = mbean_properties.get("type", mbean_domain.split(".")[1])
+        else:
+            assert 0, "Should never be here"
+
+        if mbean_properties:
+            tags += " " + " ".join(k + "=" + v for k, v in
+                                   mbean_properties.iteritems())
+
+        jmx_service = JmxMonitor.SHORT_SERVICE_NAMES.get(jmx_service, jmx_service)
+        metric = mbean_domain + "." + jmx_service.lower() + "." + metric
+
+        self.emit(metric, timestamp, value, tags)
+
 
-def kill(proc):
-  """Kills the subprocess given in argument."""
-  # Clean up after ourselves.
-  proc.stdout.close()
-  rv = proc.poll()
-  if rv is None:
-      os.kill(proc.pid, 15)
-      rv = proc.poll()
-      if rv is None:
-          os.kill(proc.pid, 9)  # Bang bang!
-          rv = proc.wait()  # This shouldn't block too long.
-  print >>sys.stderr, "warning: proc exited %d" % rv
-  return rv
-
-
-def do_on_signal(signum, func, *args, **kwargs):
```

Comment:
looks like we took this thing from a pull request and it's a bit of a mess. Since we're doing pretty much the same thing as we're doing for the flume monitor, I pulled the common logic (most of it) into a base class